src/Features/CSharp/Portable/ConvertLinq/ConvertForEachToLinqQuery/AbstractToMethodConverter.cs

```diff
@@ -0,0 +1,203 @@
+        {
+            var queryExpression = CreateQueryExpression(_selectExpression, Enumerable.Empty<SyntaxToken>(), Enumerable.Empty<SyntaxToken>());
+            var previous = FindPreviousStatementInBlock(ForEachInfo.ForEachStatement);
+
+            if (previous != null && !previous.ContainsDirectives)
+            {
+                switch (previous.Kind())
+                {
+                    case SyntaxKind.LocalDeclarationStatement:
+                        var variables = ((LocalDeclarationStatementSyntax)previous).Declaration.Variables;
+                        var lastDeclaration = variables.Last();
+                        // Check if 
+                        // var ...., list = new List<T>(); or var ..., counter = 0;
+                        // is just before the foreach.
+                        // If so, join the declaration with the query.
+                        if (_modifyingExpression is IdentifierNameSyntax identifierName &&
+                            lastDeclaration.Identifier.ValueText.Equals(identifierName.Identifier.ValueText) &&
+                            CanReplaceInitialization(lastDeclaration.Initializer.Value, semanticModel, cancellationToken))
+                        {
+                            Convert(lastDeclaration.Initializer.Value, variables.Count == 1 ? (SyntaxNode)previous : lastDeclaration);
+                            return;
+                        }
+
+                        break;
+
+                    case SyntaxKind.ExpressionStatement:
+                        // Check if 
+                        // list = new List<T>(); or counter = 0;
+                        // is just before the foreach.
+                        // If so, join the assignment with the query.
+                        if (((ExpressionStatementSyntax)previous).Expression is AssignmentExpressionSyntax assignmentExpression &&
+                            SymbolEquivalenceComparer.Instance.Equals(
+                                semanticModel.GetSymbolInfo(assignmentExpression.Left, cancellationToken).Symbol,
+                                semanticModel.GetSymbolInfo(_modifyingExpression, cancellationToken).Symbol) &&
+                            CanReplaceInitialization(assignmentExpression.Right, semanticModel, cancellationToken))
+                        {
+                            Convert(assignmentExpression.Right, previous);
+                            return;
+                        }
+
+                        break;
+                }
+            }
+
+            // At least, we already can convert to 
+            // list.AddRange(query) or counter += query.Count();
+            editor.ReplaceNode(ForEachInfo.ForEachStatement, CreateDefaultStatement(queryExpression, _modifyingExpression).WithAdditionalAnnotations(Formatter.Annotation));
+
+            void Convert(ExpressionSyntax replacingExpression, SyntaxNode nodeToRemoveIfFollowedByReturn)
```

Comment:
nit: can you place a "return;" statement above this.  It helps make it *very clear* that there is no more 'imperative executable' code in this mtehod that i have to look for.